doc/musig-reference.py

```diff
@@ -0,0 +1,453 @@
+    if is_xonly_i and not has_even_y(Q):
+        g = n - 1
+    else:
+        g = 1
+    t_i = int_from_bytes(tweak_i)
+    assert t_i < n
+    Q_i = point_add(point_mul(Q, g), point_mul(G, t_i))
+    assert not is_infinite(Q_i)
+    gacc_i = g * gacc % n
+    tacc_i = (t_i + g * tacc) % n
+    return Q_i, gacc_i, tacc_i
+
+def bytes_xor(a: bytes, b: bytes) -> bytes:
+    return bytes(x ^ y for x, y in zip(a, b))
+
+def nonce_hash(rand: bytes, aggpk: bytes, i: int, msg: bytes, extra_in: bytes) -> bytes:
+    buf = b''
+    buf += rand
+    buf += len(aggpk).to_bytes(1, 'big')
+    buf += aggpk
+    buf += i.to_bytes(1, 'big')
+    buf += len(msg).to_bytes(1, 'big')
+    buf += msg
+    buf += len(extra_in).to_bytes(4, 'big')
+    buf += extra_in
+    return int_from_bytes(tagged_hash('MuSig/nonce', buf))
+
+def nonce_gen(sk: bytes, aggpk: bytes, msg: bytes, extra_in: bytes) -> Tuple[bytes, bytes]:
+    assert len(sk) in (0, 32)
+    assert len(aggpk) in (0, 32)
+    assert len(msg) in (0, 32)
+    rand_ = secrets.token_bytes(32)
+    if len(sk) > 0:
+        rand = bytes_xor(sk, tagged_hash('MuSig/aux', rand_))
+    else:
+        rand = rand_
+    k_1 = nonce_hash(rand, aggpk, 1, msg, extra_in)
+    k_2 = nonce_hash(rand, aggpk, 2, msg, extra_in)
+    assert k_1 != 0
+    assert k_2 != 0
+    R_1_ = point_mul(G, k_1)
+    R_2_ = point_mul(G, k_2)
+    pubnonce = cbytes(R_1_) + cbytes(R_2_)
+    secnonce = bytes_from_int(k_1) + bytes_from_int(k_2)
+    return secnonce, pubnonce
+
+def nonce_agg(pubnonces: List[bytes]) -> bytes:
+    u = len(pubnonces)
+    aggnonce = b''
+    for i in (1, 2):
+        R_i_ = infinity
+        for j in range(u):
+            R_i_ = point_add(R_i_, pointc(pubnonces[j][(i-1)*33:i*33]))
+        R_i = R_i_ if not is_infinite(R_i_) else G
+        aggnonce += cbytes(R_i)
+    return aggnonce
+
+SessionContext = namedtuple('SessionContext', ['aggnonce', 'pubkeys', 'tweaks', 'is_xonly', 'msg'])
+
+def get_session_values(session_ctx: SessionContext) -> tuple[bytes, List[bytes], bytes]:
+    (aggnonce, pubkeys, tweaks, is_xonly, msg) = session_ctx
+    Q, gacc_v, tacc_v = key_agg_internal(pubkeys, tweaks, is_xonly)
+    b = int_from_bytes(tagged_hash('MuSig/noncecoef', aggnonce + bytes_from_point(Q) + msg)) % n
+    R_1 = pointc(aggnonce[0:33])
+    R_2 = pointc(aggnonce[33:66])
+    R = point_add(R_1, point_mul(R_2, b))
+    assert not is_infinite(R)
+    e = int_from_bytes(tagged_hash('BIP0340/challenge', bytes_from_point(R) + bytes_from_point(Q) + msg)) % n
+    return (Q, gacc_v, tacc_v, b, R, e)
+
+def get_session_key_agg_coeff(session_ctx: SessionContext, P: Point) -> int:
+    (_, pubkeys, _, _, _) = session_ctx
+    return key_agg_coeff(pubkeys, bytes_from_point(P))
+
+def sign(secnonce: bytes, sk: bytes, session_ctx: SessionContext) -> bytes:
+    (Q, gacc_v, _, b, R, e) = get_session_values(session_ctx)
+    k_1_ = int_from_bytes(secnonce[0:32])
+    k_2_ = int_from_bytes(secnonce[32:64])
```

Comment:
It would be nice if we could overwrite the secnonce here and add a comment that doing this is a recommendation from the BIP.

Reply:
Good idea, done!

Reply:
Good idea, though it's not possible to do this inside the function since `secnonce` is a `bytes` object which is immutable. I added some additional comments in surrounding context though.